libsolidity/analysis/StaticAnalyzer.cpp

```diff
@@ -231,6 +232,51 @@ bool StaticAnalyzer::visit(InlineAssembly const& _inlineAssembly)
 	return true;
 }
 
+bool StaticAnalyzer::visit(BinaryOperation const& _operation)
+{
+	if (
+		_operation.rightExpression().annotation().isConstant &&
+		(_operation.getOperator() == Token::Div || _operation.getOperator() == Token::Mod)
+	)
+		if (auto rhs = dynamic_pointer_cast<RationalNumberType const>(
+			ConstantEvaluator(m_errorReporter).evaluate(_operation.rightExpression())
+		))
+			if (rhs->literalValue(nullptr) == u256(0))
+				m_errorReporter.typeError(
+					_operation.location(),
+					(_operation.getOperator() == Token::Div) ? "Division by zero." : "Modulo zero."
+				);
+
+	return true;
+}
+
+bool StaticAnalyzer::visit(FunctionCall const& _functionCall)
+{
+	if (_functionCall.annotation().kind == FunctionCallKind::FunctionCall)
+	{
+		auto functionType = dynamic_pointer_cast<FunctionType const>(_functionCall.expression().annotation().type);
+		if (functionType->kind() == FunctionType::Kind::AddMod || functionType->kind() == FunctionType::Kind::MulMod)
+		{
+			solAssert(_functionCall.arguments().size() == 3, "");
+			if (
+				_functionCall.arguments()[2]->annotation().isConstant ||
+				_functionCall.arguments()[2]->annotation().type->category() == Type::Category::RationalNumber
```

Comment:
I think this line is duplicate with the dynamic cast below.

Reply:
Actually it's not - it's an *or*, not and *and* here and the cast below acts on the result of ``ConstantEvaluator``, not the argument itself.

The reason is that for example for a literal zero the ``isConstant`` annotation won't be set, but I still want to catch those. In this case ``ConstantEvaluator`` will basically return immediately, so I treat both cases the same below, in order not to have to duplicate the error message. Basically this ``if`` checks, whether one can expect, that it makes sense to invoke ``ConstantEvaluator`` on the argument below.

For division and modulo above this is not necessary, since if the right hand side is a literal, the error will already have been caught in the type checker.

Reply:
Ah right, was too late to look at it.

Would it make sense setting `isConstant` true on literals?

Reply:
I was wondering why ``isConstant`` is false for literals myself, but I haven't gotten around to look into it so far, so at this time I'm not sure whether there's a good reason not to set it to true...

Reply:
I'm fine merging this and taking care of that question in a second PR.